test/rules/invalid_case_patterns_test.dart

```diff
@@ -0,0 +1,298 @@
+
+  test_binaryExpression_logicalAnd() async {
+    await assertDiagnostics(r'''
+f(bool b) {
+  switch (b) {
+    case true && false:
+  }
+}
+''', [
+      lint(36, 13),
+    ]);
+  }
+
+  test_binaryExpression_plus() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case 1 + 2:
+  }
+}
+''', [
+      lint(43, 5),
+    ]);
+  }
+
+  test_conditionalExpression() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case true ? 1 : 2: 
+  }
+}
+''', [
+      lint(43, 12),
+      error(HintCode.DEAD_CODE, 54, 1),
+    ]);
+  }
+
+  test_constConstructorCall() async {
+    await assertDiagnostics(r'''
+class C {
+  const C();
+}
+
+f(C c) {
+  switch (c) {
+    case C():
+  }
+}
+''', [
+      lint(59, 3),
+    ]);
+  }
+
+  test_constConstructorCall_explicitConst_ok() async {
+    await assertNoDiagnostics(r'''
+class C {
+  const C();
+}
+f(C c) {
+  switch (c) {
+    case const C():
+  }
+}
+''');
+  }
+
+  test_identicalCall() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case identical(1, 2):
+  }
+}
+''', [
+      lint(43, 15),
+    ]);
+  }
+
+  test_isExpression() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case 1 is int:
+  }
+}
+''', [
+      error(HintCode.UNNECESSARY_TYPE_CHECK_TRUE, 43, 8),
+      lint(43, 8),
+    ]);
+  }
+
+  test_isNotExpression() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case 1 is! int:
+  }
+}
+''', [
+      error(HintCode.UNNECESSARY_TYPE_CHECK_FALSE, 43, 9),
+      lint(43, 9),
+    ]);
+  }
+
+  test_lengthCall() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case ''.length: 
+  }
+}
+''', [
+      lint(43, 9),
+    ]);
+  }
+
+  test_listLiteral() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case [1, 2]:
+  }
+}
+''', [
+      lint(43, 6),
+    ]);
+  }
+
+  test_listLiteral_ok() async {
+    await assertNoDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case const [1, 2]:
+  }
+}
+''');
+  }
+
+  test_listLiteral_typeArgs() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case <int>[1, 2]:
+  }
+}
+''', [
+      lint(43, 11),
+    ]);
+  }
+
+  test_mapLiteral() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+   case {'k': 'v'}:
+  }
+}
+''', [
+      lint(42, 10),
+    ]);
+  }
+
+  test_mapLiteral_ok() async {
+    await assertNoDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+   case const {'k': 'v'}:
+  }
+}
+''');
+  }
+
+  test_mapLiteral_typeArgs() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+   case <String,String>{'k': 'v'}:
+  }
+}
+''', [
+      lint(42, 25),
+    ]);
+  }
+
+  test_parenthesizedExpression() async {
+    await assertDiagnostics(r'''
+void f(Object o) {
+  switch (o) {
+    case (1):
```

Comment:
On further though, this should be OK. We do have parenthesized patterns too, so as long as the inner expression is a valid constant/pattern, the parentheses aren't a problem. You can remove this from the lint.

Reply:
Got it.  Thanks!